### PR TITLE
Implemented bluetooth support in initrd.

### PR DESCRIPTION
**Pull request by adamoa**
- Included a bluetooth module that install modules, firmwares, udev rules and bluetoothd into initrd.
- Included a dbus module as bluetoothd requires it.
- This module depends on systemd.

This is a proposed solution for the issue #270 .

### CODE REVIEW

**Comment by irreleph4nt** [comment]
Can someone please have a look at this so it can be merged? Bluetooth in initrd is very much a useful feature. Thank you.

**Comment by sigboe** [comment]
Please merge this if possible. Most bluetooth keyboard doesn't even have a USB chord option.

**Comment by hadess** [comment]
> Please merge this if possible

There are conflicts in one file, and the last 2 commits failed the test suite. It might also be good to squash the "typo fixes" commits together.

**Comment by irreleph4nt** [comment]
> > Please merge this if possible
> 
> There are conflicts in one file, and the last 2 commits failed the test suite. It might also be good to squash the "typo fixes" commits together.

Hi. What needs to happen to resolve the dracut.spec conflict?

**Comment by irreleph4nt** [comment]
I'll start tagging people now:

@lnykryn Can you please let people know what's needed to move this feature forward?

**Comment by lnykryn** [comment]
This will not work on the fedora since now we are using dbus-broker, while this commit seems to use dbus-daemon. Also could you please rebase your branch a bit, so it contains two commits? One for adding the dbus module and the second one adding the Bluetooth module.

**Comment by irreleph4nt** [comment]
> Also another thing I don't like is that after you pair a new keyboard you have to regenerate initramdisk to get there the /var/lib/bluetooth

I have created this functionality for Arch Linux's mkinitcpio, so have a bit of background. So, if you would not want to rebuild the initramfs, please let us know that how from your perspective we are meant to get the relevant data into the initramfs to actually work with a BT keyboard. Certainly you would not want people to pair a new device during boot, right? Because these changes AFAIK do not survive a reboot and would require an additional keyboard connected at the same time.

**Comment by lnykryn** [comment]
Maybe it could be stored in efivars?

**Comment by hadess** [comment]
The "right way" to do this, and the way it's implemented in devices like the Mac desktops, Lenovo all-in-ones, and plenty of other devices, is to have the linkkeys be stored on the HCI device itself. 

During boot, including EFI setup, grub, etc., the device is in HID mode, and uses the linkkeys stored on the device to talk to the mouse and keyboard. Later during boot, once storage is available, the device is switched to HCI mode, and acts as a "normal" Bluetooth adapter.

This isn't implemented in Linux because we don't have the specs for it, and haven't reverse engineered it either. The binary used to switch between the 2 modes is unsurprisingly called `hid2hci`.

**Comment by haraldh** [comment]
> Can we agree on 1) here, please?

agreed.

One more thing though... can we dynamically detect the bluetooth keyboard and only then include the bluetooth module, which then pulls in dbus?

**Comment by irreleph4nt** [comment]
> > Can we agree on 1) here, please?
> 
> agreed.
> 
> One more thing though... can we dynamically detect the bluetooth keyboard and only then include the bluetooth module, which then pulls in dbus?

You mean only pull in any Bluetooth device profile that has an identifier for keyboard and also only pull in any other data in case such a profile exists? I'd see two issues with that:
1. What if the profile is lacking that identifier?
2. Can you connect to more than one BT keyboard at once during init? I.e. what if I have paired two BT keyboards to my PC at the same time?

On question two, the way I solved this for Arch Linux was to have the user specify one BT address they would connect to during boot, because otherwise the controller might get enabled but no connection was established. I am not sure whether this would be different when using dracut?

**Comment by haraldh** [comment]
I mean, only install the dracut bluetooth + dbus modules, if there is the need for it.
The [check()](https://github.com/dracutdevs/dracut/pull/544/files#diff-816d9e86ad011566e3cd10d6d3ae25cfR4) function in the dracut bluetooth module should check, in case of hostonly, if there is an actual bluetooth connection to a keyboard.

**Comment by irreleph4nt** [comment]
> I mean, only install the dracut bluetooth + dbus modules, if there is the need for it.
> The [check()](https://github.com/dracutdevs/dracut/pull/544/files#diff-816d9e86ad011566e3cd10d6d3ae25cfR4) function in the dracut bluetooth module should check, in case of hostonly, if there is an actual bluetooth connection to a keyboard.

That particular thing I'd be unable to do. So will require someone else to look at it.

**Comment by bam80** [comment]
The feature is still wanted. Please.

**Comment by bgmilne** [comment]
> The check() function in the dracut bluetooth module should check, in case of hostonly, if there is an actual bluetooth connection to a keyboard.

As far as I know, something like this should be sufficient:
```
grep -iE 'Class=0x[0-9a-f]{3}5[4c]0' /var/lib/bluetooth/*/*/info
```

I have a Keychron K2 Bluetooth keyboard, which has:

```
# cat  /var/lib/bluetooth/*/DC\:*/info |grep -v ^Key
[General]
Name=Keychron K2
Class=0x000540
SupportedTechnologies=BR/EDR;
Trusted=true
Blocked=false
Services=00001000-0000-1000-8000-00805f9b34fb;00001124-0000-1000-8000-00805f9b34fb;00001200-0000-1000-8000-00805f9b34fb;

[LinkKey]
Type=4
PINLength=0

[DeviceID]
Source=2
Vendor=1452
Product=592
Version=283
```

Using an online [class calculator](http://bluetooth-pentest.narod.ru/software/bluetooth_class_of_device-service_generator.html), 
* a device with major class of 'peripheral', minor class of 'keyboard' should have a (hex) class of 0x540, which matches what I see, 
* toggling all of the Major Service Class checkboxes changes the leading 3 hex characters
* there is an additional minor device class of 'combo keyboard/pointing device' which uses lower 3 hex digits 0x5C0

With these conditions, a simple regex can exactly match all of the possible combinations, and we don't need to convert to binary and bitmask the values to find a device that is a keyboard.

There is the case of a computer that has previously had a bluetooth keyboard connected, but no longer has it, but I don't know how to reason about that vs. an initial ram-disk that is generated while the (only) bluetooth keyboard is not currently connected (e.g. user AFK during security updates and keyboard entered power saving).

If no-one else can implement this, I can take a look in a few more days (and also test with a different model Bluetooth keyboard).

**Comment by bgmilne** [comment]
> I mean, only install the dracut bluetooth + dbus modules, if there is the need for it.
> The [check()](https://github.com/dracutdevs/dracut/pull/544/files#diff-816d9e86ad011566e3cd10d6d3ae25cfR4) function in the dracut bluetooth module should check, in case of hostonly, if there is an actual bluetooth connection to a keyboard.

The PR as it is has check() return 255, so as far as I understand, this should only include the bluetooth support files in the ramdisk if bluetooth was requested (e.g. by `add_dracutmodules+=bluetooth`).

So I am going to make it return 0 in the case of a keyboard device having been previously used/paired, but still return 255 in other cases.

Is this what you intended ?

File: modules.d/62bluetooth/module-setup.sh [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    /usr/lib/firmware/nokia/bcmfw.bin \
+    /usr/lib/firmware/nokia/ti1273.bin
+}
+
+install() {
+  inst_multiple \
+    /usr/libexec/bluetooth/bluetoothd \
+    /usr/lib/systemd/system/bluetooth.target \
+    /usr/lib/systemd/system/bluetooth.service \
+    /etc/bluetooth/main.conf \
```
Comment (haraldh) [inline modules.d/62bluetooth/module-setup.sh]: What about other config files:
```console
$ strings /usr/libexec/bluetooth/bluetoothd|fgrep etc/bluetooth
/etc/bluetooth/network.conf
/etc/bluetooth/input.conf
/etc/bluetooth/main.conf
```

**Comment by haraldh** [comment]
> > I mean, only install the dracut bluetooth + dbus modules, if there is the need for it.
> > The [check()](https://github.com/dracutdevs/dracut/pull/544/files#diff-816d9e86ad011566e3cd10d6d3ae25cfR4) function in the dracut bluetooth module should check, in case of hostonly, if there is an actual bluetooth connection to a keyboard.
> 
> The PR as it is has check() return 255, so as far as I understand, this should only include the bluetooth support files in the ramdisk if bluetooth was requested (e.g. by `add_dracutmodules+=bluetooth`).
> 
> So I am going to make it return 0 in the case of a keyboard device having been previously used/paired, but still return 255 in other cases.
> 
> Is this what you intended ?

yes

**Comment by haraldh** [comment]
and please squash your commits into two. one for dbus and one for bluetooth

**Comment by irreleph4nt** [comment]
@bgmilne & @haraldh 
Are we making progress on getting bluetooth added to dracut? If this stalled again, then what is the current holdup, please?

**Comment by bgmilne** [comment]
> @bgmilne & @haraldh
> Are we making progress on getting bluetooth added to dracut? If this stalled again, then what is the current holdup, please?

I have tested the changes I proposed above, and they seem to work correctly from the perspective of only including the additional files when a Bluetooth keyboard has been paired/used on a system before.

However, the distribution I use on my physical desktop machine had configured dracut to not enable systemd in the initrd, and I have not been able to get my system to enable my keyboard during boot yet (but, due to work demands, I wasn't able to spend too much time on that).

I will send/update a PR over the weekend with what I have, and maybe someone else who was previously able to get a keyboard working can test that the final changes still work.

**Comment by irreleph4nt** [comment]
@bgmilne Have you had a chance to revisit your code yet?

File: modules.d/62bluetooth/module-setup.sh (hunk 1 above)
Comment (irreleph4nt) [reply modules.d/62bluetooth/module-setup.sh:42]: @haraldh Those files are not required for bluetooth to work. What's needed is 3 things:
 i) main.conf from bluez, 
ii) relevant files (like firmware, software, modules) to enable BT generally, which includes bluez and possibly d-bus, 
iii) the device profile so bluez actually connects to something during boot
All 3 are already present in this PR.

**Comment by irreleph4nt** [comment]
@bgmilne Kind reminder. Have you had a chance to look at this again? If not, what's the particular issue you are facing? Sounds like that with systemd enabled, no BT connectivity is established at boot?

**Comment by danimo** [comment]
I have landed the dbus module. Feel free to submit improvements along with the bluetooth module.

**Comment by irreleph4nt** [comment]
> I have landed the dbus module. Feel free to submit improvements along with the bluetooth module.

https://github.com/dracutdevs/dracut/blob/6fee16eab0430a38ca8b8845d9198f9dc86ad119/modules.d/06dbus/module-setup.sh#L64

Am I understanding the above code line correctly that this would get us out of the problem that bluetooth for us always started right **_after_** you had to enter the password for your encrypted drives? That has been the only real pain point with BT in dracut, as due to it BT was essentially useless.

**Comment by danimo** [comment]
> Am I understanding the above code line correctly that this would get us out of the problem that bluetooth for us always started right **_after_** you had to enter the password for your encrypted drives? That has been the only real pain point with BT in dracut, as due to it BT was essentially useless.

It should do that. If not, it's unexpected and we need to fix it.

**Comment by sigboe** [comment]
I don't want this merge request to go stale

**Comment by haraldh** [comment]
still needs to be rebased

**Comment by jo5ef** [comment]
I'm also hoping this will be merged.

**Comment by sigboe** [comment]
I ended up using clevis to store my luks key in the tmp chip, then you will have to trust your display manager, but this may be good enough depending on what work you do. Ask the CSI at your work if *you* can trust your display manager for login. Assuming he is competent of course 😄 

https://aboutcher.co.uk/2020/06/fedora-linux-luks-encryption-with-tpm-unlock/

**Comment by NaruFGT** [comment]
I'm interested in the functionality this provides.

**Comment by johannbg** [comment]
Closing this stale PR since BT support has been added via #1139

**Comment by tunix** [comment]
Hi! I've Fedora Silverblue 35 running on my computer and have the latest updates installed. I tried to enable bluetooth module by adding a file like this:

```
# cat /etc/dracut.conf.d/ble.conf
add_dracutmodules+=" bluetooth "
```

And then tried to regenerate initramfs with the following command:

```
dracut --regenerate-all --force
```

I can see the bluetooth related files inside initramfs including the files that are probably needed for my keyboard:

```
# lsinitrd initramfs-5.16.12-200.fc35.x86_64.img |grep bluetooth
bluetooth
drwxr-xr-x   1 root     root            0 Jan  1  1970 etc/bluetooth
-rw-r--r--   1 root     root        10841 Jan  1  1970 etc/bluetooth/main.conf
-rw-r--r--   1 root     root         1182 Jan  1  1970 etc/dbus-1/system.d/bluetooth.conf
drwxr-xr-x   1 root     root            0 Jan  1  1970 etc/systemd/system/bluetooth.target.wants
lrwxrwxrwx   1 root     root           41 Jan  1  1970 etc/systemd/system/bluetooth.target.wants/bluetooth.service -> /usr/lib/systemd/system/bluetooth.service
lrwxrwxrwx   1 root     root           41 Jan  1  1970 etc/systemd/system/dbus-org.bluez.service -> /usr/lib/systemd/system/bluetooth.service
-rwxr-xr-x   1 root     root       288200 Jan  1  1970 usr/bin/bluetoothctl
drwxr-xr-x   1 root     root            0 Jan  1  1970 usr/libexec/bluetooth
-rwxr-xr-x   1 root     root      1241576 Jan  1  1970 usr/libexec/bluetooth/bluetoothd
drwxr-xr-x   1 root     root            0 Jan  1  1970 usr/lib/modules/5.16.12-200.fc35.x86_64/kernel/drivers/bluetooth
-rw-r--r--   1 root     root        26680 Jan  1  1970 usr/lib/modules/5.16.12-200.fc35.x86_64/kernel/drivers/bluetooth/btbcm.ko.xz
-rw-r--r--   1 root     root        36504 Jan  1  1970 usr/lib/modules/5.16.12-200.fc35.x86_64/kernel/drivers/bluetooth/btintel.ko.xz
-rw-r--r--   1 root     root        29044 Jan  1  1970 usr/lib/modules/5.16.12-200.fc35.x86_64/kernel/drivers/bluetooth/btrtl.ko.xz
-rw-r--r--   1 root     root        46788 Jan  1  1970 usr/lib/modules/5.16.12-200.fc35.x86_64/kernel/drivers/bluetooth/btusb.ko.xz
drwxr-xr-x   1 root     root            0 Jan  1  1970 usr/lib/modules/5.16.12-200.fc35.x86_64/kernel/net/bluetooth
-rw-r--r--   1 root     root       281012 Jan  1  1970 usr/lib/modules/5.16.12-200.fc35.x86_64/kernel/net/bluetooth/bluetooth.ko.xz
drwxr-xr-x   1 root     root            0 Jan  1  1970 usr/lib/modules/5.16.12-200.fc35.x86_64/kernel/net/bluetooth/bnep
-rw-r--r--   1 root     root        33300 Jan  1  1970 usr/lib/modules/5.16.12-200.fc35.x86_64/kernel/net/bluetooth/bnep/bnep.ko.xz
drwxr-xr-x   1 root     root            0 Jan  1  1970 usr/lib/modules/5.16.12-200.fc35.x86_64/kernel/net/bluetooth/hidp
-rw-r--r--   1 root     root        37252 Jan  1  1970 usr/lib/modules/5.16.12-200.fc35.x86_64/kernel/net/bluetooth/hidp/hidp.ko.xz
drwxr-xr-x   1 root     root            0 Jan  1  1970 usr/lib/modules/5.16.12-200.fc35.x86_64/kernel/net/bluetooth/rfcomm
-rw-r--r--   1 root     root        61760 Jan  1  1970 usr/lib/modules/5.16.12-200.fc35.x86_64/kernel/net/bluetooth/rfcomm/rfcomm.ko.xz
-rw-r--r--   1 root     root          822 Jan  1  1970 usr/lib/systemd/system/bluetooth.service
-rw-r--r--   1 root     root          435 Jan  1  1970 usr/lib/systemd/system/bluetooth.target
drwxr-xr-x   1 root     root            0 Jan  1  1970 var/lib/bluetooth
drwxr-xr-x   1 root     root            0 Jan  1  1970 var/lib/bluetooth/58:96:1D:2C:14:4B
drwxr-xr-x   1 root     root            0 Jan  1  1970 var/lib/bluetooth/58:96:1D:2C:14:4B/34:88:5D:A3:D1:A7
-rw-------   1 root     root          367 Jan  1  1970 var/lib/bluetooth/58:96:1D:2C:14:4B/34:88:5D:A3:D1:A7/info
....
drwxr-xr-x   1 root     root            0 Jan  1  1970 var/lib/bluetooth/mesh
```

But when I reboot my computer, it doesn't get paired with the keyboard at LUKS prompt so I can't type my encryption password in. Am I doing something wrong? Do I need to do additional stuff?

**Pull request closed without merging**